### PR TITLE
Replace EngineInterface (from FrameworkBundle) by Twig\Environment

### LINKED ISSUE #1364: User Deprecated: The "templating.locator" service is deprecated
There is a conflict with `User Deprecated: The "templating.locator" service is deprecated`:
https://github.com/sonata-project/SonataPageBundle/blob/b1af3988017b32a2ea58fcfd61ae90ea019cee17/src/Block/BreadcrumbBlockService.php#L24
In all services with extended BreadcrumbBlockService we can't change from Symfony\Bundle\FrameworkBundle\Templating\EngineInterface to Twig\Environment

### PR DESCRIPTION
**Pull request by Daric971**
## Subject
I'm targeting this branch, because [4.x](https://github.com/sonata-project/SonataPageBundle/tree/4.x)  will be compatible with Symfony 5

Closes #1364.

## Changelog
 - Removed usage of deprecated `EngineInterface` class

### CODE REVIEW

File: src/Block/PageListBlockService.php [hunk 1]
```diff
@@ -15,22 +15,22 @@
...
 use Symfony\Component\Form\Extension\Core\Type\TextType;
 use Symfony\Component\HttpFoundation\Response;
 use Symfony\Component\OptionsResolver\OptionsResolver;
+use Twig\Environment;
 
 /**
  * @final since sonata-project/page-bundle 3.26
  */
-class PageListBlockService extends AbstractAdminBlockService
+class PageListBlockService extends AbstractBlockService
```
Comment (eerison) [inline src/Block/PageListBlockService.php]: can you add this class as `final`, and remove the annotation!

File: src/Block/BreadcrumbBlockService.php [hunk 2]
```diff
@@ -41,7 +41,7 @@ class BreadcrumbBlockService extends BaseBreadcrumbMenuBlockService
      * @param string $context
      * @param string $name
      */
-    public function __construct($context, $name, EngineInterface $templating, MenuProviderInterface $menuProvider, FactoryInterface $factory, CmsManagerSelectorInterface $cmsSelector)
+    public function __construct($context, $name, Environment $templating, MenuProviderInterface $menuProvider, FactoryInterface $factory, CmsManagerSelectorInterface $cmsSelector)
```
Comment (eerison) [inline src/Block/BreadcrumbBlockService.php]: @VincentLanglet, this class should be final?

other question, @Daric971 should provider some BC compatibility?

File: src/Page/TemplateManager.php [hunk 3]
```diff
@@ -29,7 +32,7 @@ class TemplateManager implements TemplateManagerInterface
     /**
      * Templating engine.
      *
-     * @var EngineInterface
+     * @var Environment
```
Comment (eerison) [inline src/Page/TemplateManager.php]: can you add final for this class too?

File: src/Page/TemplateManagerInterface.php [hunk 4]
```diff
@@ -25,14 +25,8 @@ interface TemplateManagerInterface
...
      * Renders a template code.
-     *
-     * @param string   $code       Template code
-     * @param array    $parameters An array of view parameters
-     * @param Response $response   Response to update
-     *
-     * @return Response
      */
-    public function renderResponse($code, array $parameters = [], ?Response $response = null);
+    public function renderResponse(string $code, array $parameters = []): Response;
```
Comment (eerison) [inline src/Page/TemplateManagerInterface.php]: why did you removed `$response` parameter?

File: src/Resources/config/block.xml [hunk 5]
```diff
@@ -35,7 +35,7 @@
             <tag name="sonata.breadcrumb"/>
             <argument>page</argument>
             <argument>sonata.page.block.breadcrumb</argument>
-            <argument type="service" id="sonata.templating"/>
```
Comment (eerison) [inline src/Resources/config/block.xml:38]: hmmm why are you removing `sonata.templating`, we don't need this Sonata Templating anymore?

File: tests/Block/PageListBlockServiceTest.php [hunk 6]
```diff
@@ -70,18 +70,9 @@ public function testExecute(): void
...
-        static::assertSame('@SonataPage/Block/block_pagelist.html.twig', $this->templating->view);
-
-        static::assertSame($blockContext, $this->templating->parameters['context']);
-        static::assertIsArray($this->templating->parameters['settings']);
-        static::assertInstanceOf(BlockInterface::class, $this->templating->parameters['block']);
-        static::assertCount(2, $this->templating->parameters['elements']);
-        static::assertContains($page1, $this->templating->parameters['elements']);
-        static::assertContains($page2, $this->templating->parameters['elements']);
-        static::assertCount(1, $this->templating->parameters['systemElements']);
-        static::assertContains($systemPage, $this->templating->parameters['systemElements']);
```
Comment (eerison) [inline tests/Block/PageListBlockServiceTest.php:86]: to replace `EngineInterface` to `Twig\Environment`, do you need to remove all those asserts?

File: src/Block/BreadcrumbBlockService.php (hunk 2 above)
Comment (VincentLanglet) [reply src/Block/BreadcrumbBlockService.php]: `@final since sonata-project/page-bundle 3.26` but all the final annotation can be done in another PR.

File: src/Resources/config/block.xml (hunk 5 above)
Comment (VincentLanglet) [reply src/Resources/config/block.xml:38]: The goal of the PR is to pass twig everywhere instead of the Templating.

File: tests/Block/PageListBlockServiceTest.php (hunk 6 above)
Comment (VincentLanglet) [reply tests/Block/PageListBlockServiceTest.php:86]: Can we have a better assert than just checking the return value is a response ?

File: src/Block/BreadcrumbBlockService.php (hunk 2 above)
Comment (VincentLanglet) [reply src/Block/BreadcrumbBlockService.php]: Because EngineInterface is not accepted anymore, the SEO v2 support is not possible anymore.
So you should drop the support in the composer.json.

I don't think it's worth it since the parent class already have this BC-break.

File: composer.json [hunk 7]
```diff
@@ -33,7 +33,7 @@
         "sonata-project/doctrine-orm-admin-bundle": "^3.19",
         "sonata-project/form-extensions": "^0.1.1 || ^1.4",
         "sonata-project/notification-bundle": "^3.8",
-        "sonata-project/seo-bundle": "^2.11",
+        "sonata-project/seo-bundle": "^2.11 || >=3",
```
Comment (VincentLanglet) [inline composer.json]:
```suggestion
        "sonata-project/seo-bundle": "^3.0",
```

**Review by VincentLanglet — CHANGES_REQUESTED** [review]

File: src/Page/TemplateManager.php [hunk 8]
```diff
@@ -103,12 +104,25 @@ public function getAll()
...
+     * @throws RuntimeError
+     * @throws LoaderError
+     */
+    public function renderResponse(string $code, array $parameters = [], ?Response $response = null): Response
     {
-        return $this->engine->renderResponse(
-            $this->getTemplatePath($code),
-            array_merge($this->defaultParameters, $parameters),
-            $response
+        if (null === $response) {
```
Comment (VincentLanglet) [inline src/Page/TemplateManager.php]:
```
$response ??= new Response();

return $response->setContent(
            $this->engine->render(
                $this->getTemplatePath($code),
                array_merge($this->defaultParameters, $parameters),
            )
        );
```

seems enough

**Review by VincentLanglet — CHANGES_REQUESTED** [review]

File: composer.json [hunk 9]
```diff
@@ -33,7 +33,7 @@
         "sonata-project/doctrine-orm-admin-bundle": "^3.19",
         "sonata-project/form-extensions": "^0.1.1 || ^1.4",
         "sonata-project/notification-bundle": "^3.8",
-        "sonata-project/seo-bundle": "^2.11",
+        "sonata-project/seo-bundle": "^2.11 || ^3.0",
```
Comment (VincentLanglet) [inline composer.json]:
```suggestion
        "sonata-project/seo-bundle": "^3.0",
```

File: src/Listener/ResponseListener.php [hunk 10]
```diff
@@ -17,40 +17,38 @@
...
      */
-    protected $decoratorStrategy;
+    private $decoratorStrategy;
 
     /**
-     * @var EngineInterface
+     * @var Environment
      */
-    protected $templating;
+    private $templating;
```
Comment (jordisala1991) [inline src/Listener/ResponseListener.php]: I think we should change all $templating for $environment or $twig ( I think I prefer twig but definitely not $templating )

File: composer.json [hunk 11]
```diff
@@ -21,7 +21,7 @@
     ],
     "homepage": "https://docs.sonata-project.org/projects/SonataPageBundle",
     "require": {
-        "php": "^7.3 || ^8.0",
+        "php": "^7.4 || ^8.0",
```
Comment (jordisala1991) [inline composer.json]: If templating is not needed anymore, please remove templating dependency

**Review by jordisala1991 — COMMENTED** [review]
The strict variables from twig should be true

File: src/Block/ChildrenPagesBlockService.php [hunk 12]
```diff
@@ -26,34 +26,23 @@
...
-    {
-        parent::__construct($name, $templating);
+    private SiteSelectorInterface$siteSelector;
 
+    private CmsManagerSelectorInterface $cmsManagerSelector;
+
+    public function __construct(string $name, Environment $twig, SiteSelectorInterface $siteSelector, CmsManagerSelectorInterface $cmsManagerSelector)
+    {
+        $this->name = $name;
+        $this->twig = $twig; //@phpstan-ignore-line
```
Comment (eerison) [inline src/Block/ChildrenPagesBlockService.php]: why do you need to ignore this line?

Comment (VincentLanglet) [inline src/Block/ChildrenPagesBlockService.php]:
```suggestion
        $this->twig = $twig;
```
Shouldn't be needed.

If there is an error it must be fixed.

File: src/Block/SharedBlockBlockService.php [hunk 13]
```diff
@@ -25,47 +25,35 @@
...
-     * @param string $name
-     *
      * @psalm-suppress ContainerDependency
      */
-    public function __construct($name, EngineInterface $templating, ContainerInterface $container, ManagerInterface $blockManager)
+    public function __construct(string $name, Environment $twig, ContainerInterface $container, ManagerInterface $blockManager, SharedBlockAdmin $sharedBlockAdmin)
     {
         $this->name = $name;
-        $this->templating = $templating;
+        $this->twig = $twig; //@phpstan-ignore-line
```
Comment (VincentLanglet) [inline src/Block/SharedBlockBlockService.php]:
```suggestion
        $this->twig = $twig;
```

File: tests/Block/ContainerBlockServiceTest.php [hunk 14]
```diff
@@ -73,7 +73,7 @@ public function testLayout(): void
          ]);
 
         $service->execute($blockContext);
-
+        // TODO: Make new test before when block-bundle will update
```
Comment (VincentLanglet) [inline tests/Block/ContainerBlockServiceTest.php]: I don't understand the todo

File: tests/Block/PageListBlockServiceTest.php [hunk 15]
```diff
@@ -70,18 +70,20 @@ public function testExecute(): void
...
+        $blockService = new PageListBlockService('block.service', $this->twig, $this->pageManager);
+        $response = $blockService->execute($blockContext);
+
+        static::assertInstanceOf(Response::class, $response);
+        // TODO: Make new test before when block-bundle will update
```
Comment (VincentLanglet) [inline tests/Block/PageListBlockServiceTest.php]: What's needed to do it now ?

**Review by VincentLanglet — CHANGES_REQUESTED** [review]

**Comment by VincentLanglet** [comment]
Hi @Daric971, do you have time to finish this PR ?

**Comment by Daric971** [comment]
Hi, I have a difficulty. I have a service that is not called and I cannot figure out the origin of the problem.

**Comment by eerison** [comment]
> Hi, I have a difficulty. I have a service that is not called and I cannot figure out the origin of the problem.

which service do you mean?

File: composer.json [hunk 16]
```diff
@@ -25,13 +25,13 @@
         "cocur/slugify": "^3.0 || ^4.0",
         "doctrine/doctrine-bundle": "^1.12.3 || ^2.0",
         "doctrine/persistence": "^1.3.4 || ^2.0",
-        "sonata-project/admin-bundle": "^3.99",
-        "sonata-project/block-bundle": "^3.20",
+        "sonata-project/admin-bundle": "^3.91.1",
```
Comment (VincentLanglet) [inline composer.json]:
```suggestion
        "sonata-project/admin-bundle": "^3.99",
```
This line shouldn't be reverted.

File: src/Admin/BlockAdmin.php [hunk 17]
```diff
@@ -216,14 +195,6 @@ private function configureBlockFields(FormMapper $form, BlockInterface $block):
 
         $service = $this->blockManager->get($block);
 
-        if (!$service instanceof BlockServiceInterface) {
```
Comment (VincentLanglet) [inline src/Admin/BlockAdmin.php]: Why this removal ?

File: src/Admin/BlockAdmin.php [hunk 18]
```diff
@@ -240,11 +211,7 @@ private function configureBlockFields(FormMapper $form, BlockInterface $block):
                 \E_USER_DEPRECATED
             );
 
-            if ($block->getId() > 0) {
-                $service->buildEditForm($form, $block);
-            } else {
-                $service->buildCreateForm($form, $block);
-            }
+            $service->buildEditForm($form, $block);
```
Comment (VincentLanglet) [inline src/Admin/BlockAdmin.php]: I don't think this should be changed.

File: src/Admin/SharedBlockAdmin.php [hunk 19]
```diff
@@ -98,14 +97,6 @@ private function configureBlockFields(FormMapper $form, BlockInterface $block):
 
         $service = $this->blockManager->get($block);
 
-        if (!$service instanceof BlockServiceInterface) {
```
Comment (VincentLanglet) [inline src/Admin/SharedBlockAdmin.php:101]: Why this removal ?

File: src/Admin/SharedBlockAdmin.php [hunk 20]
```diff
@@ -122,11 +113,7 @@ private function configureBlockFields(FormMapper $form, BlockInterface $block):
                 \E_USER_DEPRECATED
             );
 
-            if ($block->getId() > 0) {
-                $service->buildEditForm($form, $block);
-            } else {
-                $service->buildCreateForm($form, $block);
-            }
+            $service->buildEditForm($form, $block);
```
Comment (VincentLanglet) [inline src/Admin/SharedBlockAdmin.php:116]: Why this change ?

File: src/Form/Type/PageSelectorType.php [hunk 21]
```diff
@@ -61,7 +60,7 @@ public function configureOptions(OptionsResolver $resolver): void
         ]);
     }
 
-    public function setDefaultOptions(OptionsResolverInterface $resolver): void
+    public function setDefaultOptions(Options $resolver): void
```
Comment (VincentLanglet) [inline src/Form/Type/PageSelectorType.php:63]: isn't it supposed to be an OptionResolver ?

File: src/Site/BlockRendererInterface.php [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\PageBundle\Site;
+
+use Sonata\BlockBundle\Block\BlockRendererInterface as BaseBlockRendererInterface;
+
+interface BlockRendererInterface extends BaseBlockRendererInterface
```
Comment (VincentLanglet) [inline src/Site/BlockRendererInterface.php]: What's the purpose of this interface ?

**Review by VincentLanglet — CHANGES_REQUESTED** [review]

File: src/Form/Type/PageSelectorType.php (hunk 21 above)
Comment (VincentLanglet) [inline src/Form/Type/PageSelectorType.php]:
```suggestion
    public function setDefaultOptions(OptionsResolver $resolver): void
```

File: src/Admin/SharedBlockAdmin.php [hunk 23]
```diff
@@ -122,11 +122,7 @@ private function configureBlockFields(FormMapper $form, BlockInterface $block):
...
```
Comment (VincentLanglet) [inline src/Admin/SharedBlockAdmin.php]: I don't think this should be changed.

**Review by VincentLanglet — CHANGES_REQUESTED** [review]

**Comment by VincentLanglet** [comment]
Does it seem ok for you @eerison ?

**Review by VincentLanglet — APPROVED** [review]

**Comment by VincentLanglet** [comment]
@Hanmac If you want to help for the review ;)

**Comment by Hanmac** [comment]
> @Hanmac If you want to help for the review ;)

ugh i would try to split up the PR into smaller chunks ... like for example the Form stuff or the smaller changes on the listeners could be extra

File: src/Admin/SharedBlockAdmin.php [hunk 24]
```diff
@@ -77,8 +77,8 @@ protected function configureFormFields(FormMapper $form): void
 
         $form
             ->with('form.field_group_general')
-                ->add('name', null, ['required' => true])
-                ->add('enabled')
+            ->add('name', null, ['required' => true])
+            ->add('enabled')
```
Comment (jordisala1991) [inline src/Admin/SharedBlockAdmin.php:81]: please revert this change, indentations were okay.

File: src/Controller/PageAdminController.php [hunk 25]
```diff
@@ -139,9 +132,9 @@ public function createAction(?Request $request = null)
 
             if (1 === \count($sites)) {
                 return $this->redirect($this->admin->generateUrl('create', [
-                    'siteId' => $sites[0]->getId(),
-                    'uniqid' => $this->admin->getUniqid(),
-                ] + $request->query->all()));
+                        'siteId' => $sites[0]->getId(),
+                        'uniqid' => $this->admin->getUniqid(),
+                    ] + $request->query->all()));
```
Comment (jordisala1991) [inline src/Controller/PageAdminController.php:137]: Indentations were okay previously

File: src/DependencyInjection/SonataPageExtension.php [hunk 26]
```diff
@@ -110,8 +101,7 @@ public function load(array $configs, ContainerBuilder $container): void
         if (isset($bundles['SonataDoctrineBundle'])) {
             $this->registerSonataDoctrineMapping($config);
         } else {
-            // NEXT MAJOR: Remove this line and throw error when not registering SonataDoctrineBundle
-            $this->registerDoctrineMapping($config);
+            throw new ServiceNotFoundException('SonataDoctrineBundle');
```
Comment (jordisala1991) [inline src/DependencyInjection/SonataPageExtension.php:104]: This throw should have another message

File: tests/Admin/PageAdminTest.php [hunk 27]
```diff
@@ -40,10 +40,12 @@ public function testTabMenuHasLinksWithSubSite(): void
         $site = $this->createStub(Site::class);
         $site->method('getRelativePath')->willReturn('/my-subsite');
 
-        $page = new Page();
-        $page->setRouteName(PageInterface::PAGE_ROUTE_CMS_NAME);
-        $page->setUrl('/my-page');
-        $page->setSite($site);
+        $page = $this->createStub(PageInterface::class);
```
Comment (jordisala1991) [inline tests/Admin/PageAdminTest.php:43]: I dont think mocking things is better than not mocked code

File: tests/Block/PageListBlockServiceTest.php [hunk 28]
```diff
@@ -16,17 +16,17 @@
...
 use Sonata\PageBundle\Model\Page;
 use Sonata\PageBundle\Model\PageInterface;
 use Sonata\PageBundle\Model\PageManagerInterface;
+use Symfony\Component\HttpFoundation\Response;
 
 final class PageListBlockServiceTest extends BlockServiceTestCase
 {
     /**
-     * @var PageManagerInterface&MockObject
+     * @var MockObject|PageManagerInterface
```
Comment (jordisala1991) [inline tests/Block/PageListBlockServiceTest.php:29]: before was better

File: src/Listener/ResponseListener.php [hunk 29]
```diff
@@ -96,7 +77,7 @@ public function onCoreResponse(FilterResponseEvent $event): void
             $response->setPrivate();
 
             if (!$request->cookies->has('sonata_page_is_editor')) {
-                $response->headers->setCookie(Cookie::create('sonata_page_is_editor', '1'));
+                $response->headers->setCookie(new Cookie('sonata_page_is_editor', '1'));
```
Comment (jordisala1991) [inline src/Listener/ResponseListener.php:80]: is this change correct? cc @VincentLanglet

**Review by jordisala1991 — COMMENTED** [review]
Too many changed , hard to review, this is no longer a pr to remove usage of template component.

**Comment by Daric971** [comment]
hi @Hanmac, so, if I removed change for:
- form
- Listener
- Mapper
it'll be good?

**Comment by Hanmac** [comment]
> hi @Hanmac, so, if I removed change for:
> 
>     * form
> 
>     * Listener
> 
>     * Mapper
>       it'll be good?

maybe, i will try to make such changes in extra PR if able, you don't need to remove them for now

File: src/Route/RoutePageGenerator.php [hunk 30]
```diff
@@ -25,30 +25,16 @@
  * This is the page generator service from existing routes.
  *
  * @author Vincent Composieux <vincent.composieux@gmail.com>
- *
- * @final since sonata-project/page-bundle 3.26
  */
-class RoutePageGenerator
+final class RoutePageGenerator
 {
-    /**
```
Comment (eerison) [inline src/Route/RoutePageGenerator.php:33]: Can you please just keep the changes that are necessary to replace `EngineInterface` to `Twig\Enverioment`

I guess there are too much unnecessary changes, it is hard to review.

**Comment by eerison** [comment]
> Does it seem ok for you @eerison ?

IMO there are too much changes that are not related with replacement to `Twig\Enverioment`, it's hard to review.

the refactor should be done in others small pull requests.

File: src/Listener/ResponseListener.php (hunk 29 above)
Comment (VincentLanglet) [reply src/Listener/ResponseListener.php:80]: No it's not indeed. please revert this one.

**Comment by eerison** [comment]
is it still a valid PR?

**Comment by VincentLanglet** [comment]
Done by https://github.com/sonata-project/SonataPageBundle/pull/1444 I assume
